Dropout.lua

```diff
@@ -14,34 +14,42 @@ function Dropout:__init(p,v1,inplace)
 end
 
 function Dropout:updateOutput(input)
-   if self.inplace then 
-      self.output = input
-   else
-      self.output:resizeAs(input):copy(input)
-   end
-   if self.train then
-      self.noise:resizeAs(input)
-      self.noise:bernoulli(1-self.p)
-      if self.v2 then
-         self.noise:div(1-self.p)
+   if self.p > 0 then
+      if self.inplace then
+          self.output = input
+      else
+          self.output:resizeAs(input):copy(input)
       end
-      self.output:cmul(self.noise)
-   elseif not self.v2 then
-      self.output:mul(1-self.p)
+      if self.train then
+          self.noise:resizeAs(input)
+          self.noise:bernoulli(1-self.p)
+          if self.v2 then
+             self.noise:div(1-self.p)
+          end
+          self.output:cmul(self.noise)
+      elseif not self.v2 then
+          self.output:mul(1-self.p)
+      end
+   else
+      self.output = input
    end
    return self.output
 end
 
 function Dropout:updateGradInput(input, gradOutput)
-   if self.train then
-      if self.inplace then
-         self.gradInput = gradOutput
+   if self.p > 0 then
+      if self.train then
+         if self.inplace then
+            self.gradInput = gradOutput
+         else
+            self.gradInput:resizeAs(gradOutput):copy(gradOutput)
+         end
+         self.gradInput:cmul(self.noise) -- simply mask the gradients with the noise vector
       else
-         self.gradInput:resizeAs(gradOutput):copy(gradOutput)
+         error('backprop only defined while training')
       end
-      self.gradInput:cmul(self.noise) -- simply mask the gradients with the noise vector
    else
-      error('backprop only defined while training')
+      self.gradInput = gradOutput
```

Comment:
same comments as in updateOutput. Doesn't respect in-place / out-of-place. instead reuse the logic in line 42-46